### ISSUE TITLE
First GAN

### ISSUE THREAD

**Issue by raspstephan** [raspstephan/nwp-downscale#29]
Some things to try out @HirtM :

- [x] Train generator more
- [x] Improve discriminator
- [x] Make ResNet Blocks for the discriminator
- [ ] U-Net generator
- [x] Try bilinear upsampling instead of pixel shuffle
- [ ] Add 90 degree rotation augmentation
- [x] Conditional Discrimnator: follow Leinonen architecture

**Comment by raspstephan** [comment]
@annavaughan I am just looking through your esr_gan notebook and I am a little confused by the Generator architecture. 
![image](https://user-images.githubusercontent.com/2786927/110097625-8b94f800-7d9f-11eb-8a72-c2e7611886d3.png)
From what I understand this means that the same convolution (self.conv2), like the same weights, is used for blocks 2-7. Isn't that kind of weird?
Same for self.bn.

**Comment by raspstephan** [comment]
I am training our most advanced GAN on the first GPU VM. I created another GPU VM `nwp-downscale-gpu2`, which should work (remember, you have to add your user again). You can find the IP on the Azure portal. JIT is also enabled, so you have to request access as usual. @HirtM @annavaughan

**Comment by raspstephan** [comment]
So some first results from training the GAN with hinge loss, etc for a longer time (like 60 epochs) @HirtM 

First the settings:

![image](https://user-images.githubusercontent.com/2786927/111642335-4ab8cc80-87fe-11eb-92b3-0091020a0a7d.png)
![image](https://user-images.githubusercontent.com/2786927/111642361-51474400-87fe-11eb-896b-9c047d74a6f2.png)

I also pretained the generator using the MSE loss for two epochs.

![image](https://user-images.githubusercontent.com/2786927/111642179-28bf4a00-87fe-11eb-87d4-f72af1d920f7.png)
![image](https://user-images.githubusercontent.com/2786927/111642215-2fe65800-87fe-11eb-8f29-cd7f1a86c837.png)
![image](https://user-images.githubusercontent.com/2786927/111642244-34ab0c00-87fe-11eb-9802-57143127d572.png)

One thing I noticed is that for most batches the discriminator loss is zero. This probably happens because of the ReLU in the Hinge loss. I would like to try the regular Wasserstein loss as well to see if that would help. Otherwise it might make sense to train the generator more.

**Comment by raspstephan** [comment]
@HirtM I also made some minor changes in the 08 notebook like adding a plotting function that can be called after each epoch and saving the disc and gen. It's currently on the stephans_gan branch. There is a merge conflict which I didn't have time to fix right now.

**Comment by raspstephan** [comment]
I tried training with the Wasserstein loss in the Experiments/01-WGAN notebook on my gan branch.
![image](https://user-images.githubusercontent.com/2786927/111757783-fcf2a180-889b-11eb-9e8a-f9ed3c2dd0fe.png)

I think in the Wasserstein loss the absolute value of the discriminator loss should get smaller which it doesn't. The results do show that there is some learning going on though.

![image](https://user-images.githubusercontent.com/2786927/111757793-011ebf00-889c-11eb-96f5-526f50fc1c0c.png)
![image](https://user-images.githubusercontent.com/2786927/111757806-054adc80-889c-11eb-96c3-e7b6ca7c8dd5.png)

One thing I just realized though is that I have spectral normalization AND the gradient penalty on. Hmm...

**Comment by HirtM** [comment]
I have just looked at the "weird" structures in a bit more detail. I don't know if there is any valuable information, but I might as well document it: 
In my example at least, every fourth grid point seems to be strongly correlated: 
![image](https://user-images.githubusercontent.com/43262391/112185928-f5672b80-8c00-11eb-8cb9-0c17eba4b1ec.png)
![image](https://user-images.githubusercontent.com/43262391/112186084-16c81780-8c01-11eb-83e6-7ffa09d0cb56.png)

It might be different for other cases though. 

Also following https://medium.com/@hirotoschwert/introduction-to-deep-super-resolution-c052d84ce8cf (see e.g. Fig. 3), it also sounds reasonable to me to try and replace the pixelshuffler, as you @raspstephan  already suggested. 

Here is another explanation for checkerboard patterns related to not matching strides and filter sizes in a deconvolution, but as far as I understand it, this should not apply to us: https://towardsdatascience.com/a-comprehensive-introduction-to-different-types-of-convolutions-in-deep-learning-669281e58215

**Comment by raspstephan** [comment]
I implemented a bunch of changes to the network architecture to make it more similar to the Leinonen Paper. Well, the GAN finally does something, see here: 
![image](https://user-images.githubusercontent.com/2786927/115515115-1618b480-a285-11eb-93cf-292205071615.png)
Now, of course, I have no idea why it "works" now because there were too many changes at once. Here are the settings:

I am using a log-transform. I have a gut feeling that that made a big difference.
![image](https://user-images.githubusercontent.com/2786927/115515248-3c3e5480-a285-11eb-8420-f94efa61ea3d.png)

Different from Leinonen, I am using spectral norm in generator AND discriminator and I am not using any l2 regularization in the generator.
![image](https://user-images.githubusercontent.com/2786927/115515346-537d4200-a285-11eb-8d8f-f3f86c6a8f65.png)

I am pretraining.
![image](https://user-images.githubusercontent.com/2786927/115515384-5ed06d80-a285-11eb-9b84-3d291a12fe6b.png)

Using the Wasserstein loss with gradient penalty and 5 disc steps per gen step. I am using the l1 loss as well for the generator but it doesn't do anything, see below. 
![image](https://user-images.githubusercontent.com/2786927/115515502-7a3b7880-a285-11eb-9de0-52928030c4d4.png)

The generator loss fluctuates wildly. The disc loss does what it should do.
![image](https://user-images.githubusercontent.com/2786927/115515989-f766ed80-a285-11eb-9705-815350405f30.png)
![image](https://user-images.githubusercontent.com/2786927/115516004-fa61de00-a285-11eb-9dad-0070ebe5129b.png)
![image](https://user-images.githubusercontent.com/2786927/115516180-27ae8c00-a286-11eb-807d-d213ae1a562a.png)


While the GAN produces something interesting, it's still far away from looking realistic. The fact that the generator loss is so wild and that the L1 loss has no impact doesn't seem right. So the question is: why is the generator loss so crazy? Why is the distribution of gen_preds_fake so different from disc_preds_fake? Shouldn't they be approximately the same?

Also interesting: it takes around 3-4 epochs (2h) to see where we are going.

**Comment by raspstephan** [comment]
Two tests: 

1. No Gradient penalty (because I thought using spectral norm in both G and D already does enough). doesn't Work at all. Losses explode. Images not good.

![image](https://user-images.githubusercontent.com/2786927/115587324-d88c4980-a2cd-11eb-8bac-9f08724d0889.png)
![image](https://user-images.githubusercontent.com/2786927/115587370-e8a42900-a2cd-11eb-8635-1ce4450e4ddd.png)

2. Using hinge loss. I mean, not catastrophic but also not good.
![image](https://user-images.githubusercontent.com/2786927/115587417-f9549f00-a2cd-11eb-9b66-7199c6bcffaf.png)
![image](https://user-images.githubusercontent.com/2786927/115587464-05d8f780-a2ce-11eb-8184-a4f1e2245820.png)

**Comment by raspstephan** [comment]
I am closing this issue and opening another one.
